cache binary pattern in `eredis_parser`

e-kmisk opened the pull request:
Use the process dict as a cache to store the compiled binary pattern.
This makes binary:match/2 significantly faster for small binaries
where size is <10000.

zuiderkwast commented on include/eredis.hrl:
```diff
@@ -54,6 +54,15 @@
+-define(NL_PATTERN,
+        case get(eredis_newline_pattern) of
+                undefined ->
+                    Pattern = binary:compile_pattern(?NL),
```
It needs to be a binary. :)

```suggestion
                    Pattern = binary:compile_pattern(<<?NL>>),
```

Btw, it's risky to use variables in macros. This macro binds or matches Pattern in the scope where it's used.

I suggest instead use function and declare it as inline. See https://www.erlang.org/doc/man/compile.html#inlining

e-kmisk replied:
Checking

zuiderkwast commented on src/eredis_parser.erl:
```diff
@@ -174,6 +176,15 @@ do_parse({multibulk_continue, _RemainingItems, _Acc} = State, <<>>) ->
+        nomatch ->
+            nomatch;
+        {NewlinePos, 2} ->
+            <<FirstLine:NewlinePos/binary, ?NL, Rest/binary>> = Data,
```
Good. Btw, do you know why we don't use [binary:split/2,3](https://www.erlang.org/doc/man/binary.html#split-2)? Is it slower than match?

e-kmisk replied:
My assumption is yes, but I can test it. Reason being that `binary:split` creates a list of binaries which is then discarded anyway, which this match expression does not do (at least I assume).

e-kmisk replied:
The size checking here is not faster than just doing the binary match expression.
```
Code                                                                                                                                                                      ||        QPS       Time     Rel
runner({A, Pos, Len}) -> <<B:Pos/binary, _C:Len/binary, _D/binary>> = A, B.                                                                                                1   25136 Ki      39 ns    100%
runner({A, Pos, Len}) -> case size(A) - Pos - Len == 0 of true -> <<B:Pos/binary, _C/binary>> = A, B; _ -> <<B:Pos/binary, _C:Len/binary, _D/binary>> = A, B end.          1   20502 Ki      48 ns     81%
```


However another tiny optimization is that matching on any 2-byte middle component is faster than matching on `?NL` exactly (and it is anyway ensured by `binary:match` that we got the right thing)

```
Code                                                                                ||        QPS       Time     Rel
runner({A, Pos, Len}) -> <<B:Pos/binary, _C:Len/binary, _D/binary>> = A, B.          1   28052 Ki      35 ns    100%
runner({A, Pos, Len}) -> <<B:Pos/binary, "\r\n", _D/binary>> = A, B.                 1   27031 Ki      36 ns     96%
```

e-kmisk replied:
I will go with `binary:split`, oddly it seems the fastest and is also not ugly

```bash
BIN="<<\"$(echo -n "b"; head -c 99 /dev/urandom | base32 -w 0 |head -c 99 ; echo -n '\r\n')\">>"
./erlperf \
'runner(Bin) -> case binary:match(Bin,get(bin)) of {P, L} -> <<A:P/binary, _B:L/binary, C/binary>> = Bin, {A, C} end.' \
    --init_runner 'put(bin,binary:compile_pattern(<<"\r\n">>)), '"${BIN} ." \
'runner(Bin) -> case binary:split(Bin,get(bin)) of [A, C] -> {A, C} end.' \
    --init_runner 'put(bin,binary:compile_pattern(<<"\r\n">>)), '"${BIN} ."
```
results in
```
Code                                                                                                                         ||        QPS       Time     Rel
runner(Bin) -> case binary:split(Bin,get(bin)) of [A, C] -> {A, C} end.                                                       1   10569 Ki      94 ns    100%
runner(Bin) -> case binary:match(Bin,get(bin)) of {P, L} -> <<A:P/binary, _B:L/binary, C/binary>> = Bin, {A, C} end.          1    7922 Ki     126 ns     74%
```

zuiderkwast commented on src/eredis_parser.erl:
```diff
@@ -174,6 +176,15 @@ do_parse({multibulk_continue, _RemainingItems, _Acc} = State, <<>>) ->
     {nested, State, Data}.
 
+%% Optimize the first clause for the most common case - when Acc is empty
+split_by_newline(Acc, Data) when size(Acc) == 0 ->
```
Are you saying that `when size(Acc) == 0` is faster than a direct match on the empty binary?

```suggestion
split_by_newline(<<>>, Data) ->
```

zuiderkwast replied:
OK. It'd be good to add a code comment about this to prevent others from simplifying this in the future.

e-kmisk commented:
By the way, process dictionary is faster than `persistent_term`:
```bash
BIN="<<\"$(echo -n "b"; head -c 99 /dev/urandom | base32 -w 0 |head -c 99 ; echo -n '\r\n')\">>"
./erlperf \
'runner({Bin,GetProcDict}) -> case binary:split(Bin,GetProcDict()) of [A, C] -> {A, C} end.' \
    --init_runner 'put(bin,binary:compile_pattern(<<"\r\n">>)),'"{${BIN} , fun() -> get(bin) end}." \
'runner({Bin,GetPersTerm}) -> case binary:split(Bin,GetPersTerm()) of [A, C] -> {A, C} end.' \
    --init_runner 'persistent_term:put(my_bin,binary:compile_pattern(<<"\r\n">>)),'"{${BIN} , fun() -> persistent_term:get(my_bin) end}." \
-l 100M
```
```
Code                                                                                               ||        QPS       Time     Rel
runner({Bin,GetProcDict}) -> case binary:split(Bin,GetProcDict()) of [A, C] -> {A, C} end.          1   11321 Ki      88 ns    100%
runner({Bin,GetPersTerm}) -> case binary:split(Bin,GetPersTerm()) of [A, C] -> {A, C} end.          1   10180 Ki      98 ns     89%
```


of course with the caveats being
* for each eredis_client process the cached compiled pattern is duplicated -> more memory use (but this object seems small, probably few dozen bytes)
* each eredis client process calls `binary:compile_pattern/1` once when they parse their first query response -> slightly longer latency on very first command (latency of a `compile_pattern` seems to be ~100ns on my machine for `\r\n` as a pattern)

zuiderkwast commented:
Perhaps we shuold include code comments ore a separate markdown file in the repo, describing the erlperf menchmarks you've done here. It's important background and useful for future optimization attempts.

zuiderkwast commented:
CI: Elvis complains about code style:

```
# src/eredis_parser.erl [FAIL]
  - operator_spaces
    - Missing space right "," on line 27
make: *** [Makefile:78: elvis] Error 1
```

zuiderkwast commented on src/eredis_parser.erl:
```diff
@@ -135,7 +137,7 @@ do_parse({StateTag, Acc}, Data) when StateTag =:= bulk_size;
     %% we received a partial response and need more data
     case split_by_newline(Acc, Data) of
-        nomatch ->
+        false ->
```
Why false rather than nomatch?

e-kmisk replied:
The `nomatch` was a specific atom related to `binary:match`. Since we are no longer using `binary:match` it is misleading imo. to keep that atom.

I picked `false` because it kind of makes sense for readability -> "if split_by_newline is false then ..." reads better to me than "if split_by_newline is nomatch then ..."

zuiderkwast replied:
I think false is more of a boolean. Some newer OTP functions return `error` in this situation, such as `maps:find(Key, Map) -> {ok, Value} | error`.

zuiderkwast replied:
"if split_by_newline is false" reads like nonsense to me. :-)

e-kmisk replied:
I think I saw it first in `lists:search` but I'm fine either way.
```
search(Pred, List) -> {value, Value} | false
```

e-kmisk replied:
I personally don't like error when it's not really an "error" just incomplete data. Let's go with continue maybe?

zuiderkwast replied:
It's an error in the sense the function failed to split the data. An atom like 'notfound' is good imo but there are many variations of this. It seems {ok, Value} and {error, Reason} or just 'ok' and 'error' is the most typical erlang style nowdays. The odd stuff in the lists module has been regarded as old odd stuff. I've followed EEP discussions where this comes up, e.g. the EEP about maps.

zuiderkwast replied:
`continue` only describes what the return value is used for where it is currently called. The name and return values of a function should not be influenced by where it happens to be called. I don't like it.

I'm fine with either of these (but I prefer the first of them):

* `error` as in
    * `maps:find(Key, Map) -> {ok, Value} | error`
    * `dict:find(Key, Dict) -> {ok, Value} | error`
    * `dict:take(Key, Dict) -> {Value, Dict1} | error`
* `none` as in
    * `proplists:lookup(Key, List) -> none | tuple()`
    * `gb_trees:lookup(Key, Tree) -> none | {value, Value}`
* `nomatch` as in
    * `re:run(Subject, RE) -> {match, Captured} | nomatch`
    * `string:find(String, SearchPattern) -> [unicode:chardata()](https://www.erlang.org/doc/man/unicode.html#type-chardata) | nomatch`

But I don't like these:

* ~~`false`~~ (as in `lists:keysearch(Key, N, TupleList) -> {value, Tuple} | false`)
* ~~`continue`~~ ...

e-kmisk replied:
`nomatch` is fine if it's not only binary:match that uses it. It also yields the minimal diff.

zuiderkwast reviewed (approved):
LGTM
